simSPI/linear_simulator/modules.py

```diff
@@ -0,0 +1,397 @@
+            rot_vol_coords[:, None, None, :, :],
+            align_corners=True,
+        )
+        projection = torch.sum(
+            rot_vol.reshape(
+                batch_sz, self.config.sidelen, self.config.sidelen, self.config.sidelen
+            ),
+            dim=proj_axis,
+        )
+        projection = projection[
+            :, None, :, :
+        ]  # add a dummy channel (for consistency w/ img fmt)
+        # --> B,C,H,W
+        return projection
+
+
+"""Module that multiplies ctf to fourier transform of the projection."""
+
+
+class CTF(torch.nn.Module):
+    """Class that multiplies ctf to fourier transform of the projection.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the CTF
+
+    """
+
+    def __init__(self, config):
+        super(CTF, self).__init__()
+
+        self.config = config
+        self.wavelength = self._get_ewavelength()
+        self.frequency = 1.0 / (self.config.ctf_size * self.config.pixel_size)
+        n2 = float(self.config.ctf_size // 2)
+        ax = torch.arange(-n2, n2 + self.config.ctf_size % 2)
+        mx, my = torch.meshgrid(ax, ax)
+        self.register_buffer("r2", mx ** 2 + my ** 2)
+        self.register_buffer("r", torch.sqrt(self.r2))
+        self.register_buffer("angleFrequency", torch.atan2(my, mx))
+
+    def _get_ewavelength(self):
+        """Output the wavelength of the electron beam.
+
+        Returns
+        -------
+        wavelength: float
+            wavelength of the electron beam
+        """
+        wavelength = 12.2639 / np.sqrt(
+            self.config.kV * 1e3 + 0.97845 * self.config.kV ** 2
+        )
+        return wavelength
+
+    def get_ctf(self, ctf_params):
+        """Output the CTF for a given params.
+
+        Parameters
+        ----------
+        ctf_params: dict of type str to {tensor}
+            contains "defocusU", "defocusV", "defocusAngle",
+            all of shape (chunks,1,1,1)
+
+        Returns
+        -------
+        hFourier: torch.Tensor
+            ctf of shape (chunks,1,ctf_size,ctf_size)
+        """
+        defocusU = ctf_params["defocusU"]
+        defocusV = ctf_params["defocusV"]
+        angleAstigmatism = ctf_params["defocusAngle"]
+        elliptical = (
+            defocusV * self.r2
+            + (defocusU - defocusV)
+            * self.r2
+            * torch.cos(self.angleFrequency - angleAstigmatism) ** 2
+        )
+        defocusContribution = (
+            np.pi * self.wavelength * 1e4 * elliptical * self.frequency ** 2
+        )
+        abberationContribution = (
+            -np.pi
+            / 2.0
+            * self.config.cs
+            * (self.wavelength ** 3)
+            * 1e7
+            * self.frequency ** 4
+            * self.r2 ** 2
+        )
+
+        argument = abberationContribution + defocusContribution
+
+        hFourier = (1 - self.config.amplitude_contrast ** 2) ** 0.5 * torch.sin(
```

Comment:
do you have the sign right? the CTF is negative close to zero, and the amplitude contrast makes it (more) negative at zero frequency?

Reply:
https://www.youtube.com/watch?v=Y8wivQTJEHQ&t=1833s

Reply:
Correct, should be `-sqrt(1-a**2).sinx + a.cosx`

Reply:
@geoffwoollard Great point! The CTF is negative at 0 in most of the literature. That is why we had the negative sign in CryoGAN (see [here](https://github.com/harshit-gupta-cor/CryoGAN/blob/ba4d33471668e79ccdd0a7706ba04d7d90a4465c/Functions/FunctionsCTF.py#L110))

However, I have tried to emulate the Relion's simulator and it uses the positive sign. The code in this PR (from cryonettorch) has been thoroughly compared w.r.t Relion and is consistent with it. Below is the CTF that is outputted using relion_project function. I am attaching a directory with .py and .sh files to replicate this image. It obtains projection from relion for a cube which is one pixel wide (delta volume). The fourier transform of the output is actually the CTF multiplied by a fourier normalization factor.
<img width="401" alt="image" src="https://user-images.githubusercontent.com/18009768/144116203-c91dc7f1-bd4a-4c73-8630-3c10317632ba.png">

Reply:
@fredericpoitevin I think you meant: -sqrt(1-a**2).sinx - a.cosx 
Because then at x=0 value of CTF will be -a.cos(0)=-a=negative.

This is exactly what we had in CryoGAN and I have described above the reason why it was changed.